content/2015.09.kinto-encryption-example.rst

```diff
@@ -0,0 +1,207 @@
+* <code>() => {}</code> is similar to <code>function() {}</code>
+(although the scoping of <code>this</code> is different).
+* <code>const foo = 'bar'</code> and <code>let foo = 'bar'</code> are similar to
+<code>var foo = 'bar'</code>.
+* <code>{ a, b }</code> is similar to <code>{ a: a, b: b }</code>.
+* <code>`Hello ${foo}`</code> (note the backticks) is similar to
+<code>'Hello ' + foo</code>.
+
+First, we need a few helper functions, which will be used later on to convert
+between binary buffers and ASCII strings.
+
+.. code-block:: javascript
+  // Helper functions:
+  const rawStringToByteArray = (str) => {
+    const strLen = str.length;
+    var byteArray = new Uint8Array(strLen);
+    for (var i = 0; i < strLen; i++) {
+      byteArray[i] = str.charCodeAt(i);
+    }
+    return byteArray;
+  };
+  const base64StringToByteArray = (base64) => {
+    return rawStringToByteArray(window.atob(base64));
+  };
+  const byteArrayToBase64String = (buffer) => {
+    var bytes = new Uint8Array(buffer);
+    var binary = '';
+    for (var i=0; i<bytes.byteLength; i++) {
+        binary += String.fromCharCode(bytes[i]);
+    }
+    return window.btoa(binary);
+  };
+
+Next, we need to create a <code>RemoteTransformer</code> that can encode and decode. And for
+that, we need to generate an AES key. We will use that key to encrypt and decrypt
+the data that is passed through the <code>RemoteTransformer</code>:
+
+.. code-block:: javascript
+  // RemoteTransformer:
+  const generateAesKey = () => {
+    // See http://www.w3.org/TR/WebCryptoAPI/#examples-symmetric-encryption
+    return window.crypto.subtle.generateKey({ name: 'AES-CBC', length: 128 },
+         false, ['encrypt', 'decrypt']);
+  };
+
+  const createTransformer = (aesKey) => {
+    const encode = (record) => {
+      const cleartext = rawStringToByteArray(JSON.stringify(record));
+      const IV = window.crypto.getRandomValues(new Uint8Array(16));
+
+      return window.crypto.subtle.encrypt({ name: 'AES-CBC', iv: IV }, aesKey,
+          cleartext).then(ciphertext => {
+        return {
+          id: record.id,
+          ciphertext: byteArrayToBase64String(new Uint8Array(ciphertext)),
+          IV: byteArrayToBase64String(IV)
+        };
+      });
+    };
+
+    const decode = (record) => {
+      const ciphertext = base64StringToByteArray(record.ciphertext);
+      const IV = base64StringToByteArray(record.IV);
+
+      return crypto.subtle.decrypt({ name: 'AES-CBC', iv: IV }, aesKey,
+          ciphertext).then(recordArrayBuffer => {
+
+        return JSON.parse(String.fromCharCode.apply(null,
+            new Uint8Array(recordArrayBuffer)));
+      }, () => {
+        record.undecryptable = true;
+        return record;
+      });
+    };
+
+    return {
+      encode,
+      decode
+    };
+  };
+
+We create two Kinto instances, so that we can test syncing from one instance
+to the other, via the public Kinto demo instance on AWS. Creating multiple Kinto
```

Comment:
What is a good information page about the demo instance? I read about it somewhere in passing, but then found the credentials in the source code of some other app. :)